modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,332 @@
+//
+// The journal represents a single JSON object, containing all of the
+// contractor's persisted data. The object is serialized as an "initial
+// object" followed by a series of update sets, one per line. Each update
+// specifies a modification.
+//
+// During operation, the object is first loaded by reading the file and
+// applying each update to the initial object. It is subsequently modified by
+// appending update sets to the file, one per line. At any time, a
+// "checkpoint" may be created, which clears the journal and starts over with
+// a new initial object. This allows for compaction of the journal file.
+//
+// In the event of power failure or other serious disruption, the most recent
+// update set may be only partially written. Partially written update sets are
+// simply ignored when reading the journal.
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/persist"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+var journalMeta = persist.Metadata{
+	Header:  "Contractor Journal",
+	Version: "1.0.0",
+}
+
+// A journal is a log of updates to a JSON object.
+type journal struct {
+	f        *os.File
+	filename string
+}
+
+// update applies the updateSet atomically to j. It syncs the underlying file
+// before returning.
+func (j *journal) update(us updateSet) error {
+	if err := json.NewEncoder(j.f).Encode(us); err != nil {
+		return err
+	}
+	return j.f.Sync()
+}
+
+// Checkpoint refreshes the journal with a new initial object. It syncs the
+// underlying file before returning.
+func (j *journal) checkpoint(data contractorPersist) error {
+	// write to a new temp file
+	tmp, err := os.Create(j.filename + "_tmp")
+	if err != nil {
+		return err
+	}
+	enc := json.NewEncoder(tmp)
+	if err := enc.Encode(journalMeta); err != nil {
+		return err
+	}
+	if err := enc.Encode(data); err != nil {
+		return err
+	}
+	if err := tmp.Sync(); err != nil {
+		return err
+	}
+
+	// atomically replace the old file with the new one
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+	if err := j.f.Close(); err != nil {
+		return err
+	}
+	if err := os.Rename(tmp.Name(), j.filename); err != nil {
+		return err
+	}
+
+	// reopen the journal
+	j.f, err = os.OpenFile(j.filename, os.O_RDWR|os.O_APPEND, 0)
+	return err
+}
+
+// Close closes the underlying file.
+func (j *journal) Close() error {
+	return j.f.Close()
+}
+
+// newJournal creates a new journal, using an empty contractorPersist as the
+// initial object.
+func newJournal(filename string) (*journal, error) {
+	f, err := os.Create(filename)
+	if err != nil {
+		return nil, err
+	}
+	// write metadata
+	enc := json.NewEncoder(f)
+	if err := enc.Encode(journalMeta); err != nil {
+		return nil, err
+	}
+	// write empty contractorPersist
+	if err := enc.Encode(contractorPersist{}); err != nil {
```

Comment:
Pretty much. The only non-empty fields are the ones that get filled in by `openJournal`. I figured that it would be necessary to check for `null` whenever you open the journal, so it's redundant to initialize them in `newJournal`. I suppose it wouldn't hurt though.